crates/editor/src/element.rs

```diff
@@ -3715,137 +3832,306 @@ impl EditorElement {
+                            top: Pixels::ZERO,
+                            right: Pixels::ZERO,
+                            bottom: Pixels::ZERO,
+                            left: ScrollbarLayout::BORDER_WIDTH,
+                        },
+                        cx.theme().colors().scrollbar_thumb_border,
+                    ));
+                })
+            }
 
-                let markers = self.editor.read(cx).scrollbar_marker_state.markers.clone();
-                for marker in markers.iter().chain(&fast_markers) {
-                    let mut marker = marker.clone();
-                    marker.bounds.origin += scrollbar_layout.hitbox.origin;
-                    cx.paint_quad(marker);
-                }
+            cx.set_cursor_style(CursorStyle::Arrow, &hitbox);
 
-                cx.paint_quad(quad(
-                    thumb_bounds,
-                    Corners::default(),
-                    cx.theme().colors().scrollbar_thumb_background,
-                    Edges {
-                        top: Pixels::ZERO,
-                        right: Pixels::ZERO,
-                        bottom: Pixels::ZERO,
-                        left: ScrollbarLayout::BORDER_WIDTH,
-                    },
-                    cx.theme().colors().scrollbar_thumb_border,
-                ));
-            });
-        }
+            cx.on_mouse_event({
+                let editor = self.editor.clone();
 
-        cx.set_cursor_style(CursorStyle::Arrow, &scrollbar_layout.hitbox);
+                // there may be a way to avoid this clone
+                let hitbox = hitbox.clone();
 
-        let row_height = scrollbar_layout.row_height;
-        let row_range = scrollbar_layout.visible_row_range.clone();
+                let mut mouse_position = cx.mouse_position();
+                move |event: &MouseMoveEvent, phase, cx| {
+                    if phase == DispatchPhase::Capture {
+                        return;
+                    }
 
-        cx.on_mouse_event({
-            let editor = self.editor.clone();
-            let hitbox = scrollbar_layout.hitbox.clone();
-            let mut mouse_position = cx.mouse_position();
-            move |event: &MouseMoveEvent, phase, cx| {
-                if phase == DispatchPhase::Capture {
-                    return;
-                }
+                    editor.update(cx, |editor, cx| {
+                        if event.pressed_button == Some(MouseButton::Left)
+                            && editor
+                                .scroll_manager
+                                .is_dragging_scrollbar(Axis::Horizontal)
+                        {
+                            let x = mouse_position.x;
+                            let new_x = event.position.x;
+                            if (hitbox.left()..hitbox.right()).contains(&x) {
+                                let mut position = editor.scroll_position(cx);
+
+                                position.x += (new_x - x) / text_unit_size;
```

Comment:
 Let's use something like `checked_div` here. Better safe than sorry.

Reply:
These are `f32`. I don't see a way to use `checked_div` here. The division is by `text_unit_size` in every case. Looking closer it looks like `text_unit_size` could theoretically be 0 in some cases (only when the scrollbar thumb is the exact same size as the bar), but I've never seen it happen. For the sake of safety, I'll add a separate check for 0 in `layout_scrollbars` (where the value is set) and if it's 0 the scrollbar will not be drawn. The thumb would not be movable in that case anyway.

Reply:
I don't see a way to do this that doesn't break the functionality. This already seems to be handled elsewhere as well, since the scrollbars do not show anyway when the thumb would take up the entire bar. I can look into this more if desired.

This division has been present for months now, and I haven't touched it to avoid breaking things. I can't guarantee that `text_unit_size` will never be 0 though.

Reply:
That's fine. We can keep it as is, especially given that we're not going to do another release till early next year (which should give this code some testing time on Nightly). I suggested checking for 0 here just because there's no real precedent for it and it would make the code better IMHO, but I think we can get by.